### PR TITLE
[ROCm][Perf] Skip cleaning sparse prefill MQA logits

### PR DESCRIPTION
**Pull request by amd-sriram**
## Summary

The sparse prefill path creates the MQA-logits buffer with `torch.full(..., -inf)`.
This launches an FP32 fill before every logits kernel.

This PR passes `clean_logits=False` to AITER, so the buffer uses `torch.empty`
and the fill is removed.

## Safety

The only consumer is `top_k_per_row_prefill`. It reads each row only inside
`[cu_seqlen_ks, cu_seqlen_ke)`. Its vectorized path also guards the final partial
load, so padding values are never read. The top-k result is unchanged.

## Results

MI355X, TP=4, rank 0, comparable before/after captures:

| metric | before | after |
|---|---:|---:|
| FP32 fill launches | 18,162 | **0** |
| FP32 fill GPU time | 1,250.2 ms | **0 ms** |
| MQA-logits launches | 18,108 | 17,930 |

The removed fill was **0.89% of total GPU kernel time**. In the sample below it
takes 74.6 µs and is fully serialized before the logits kernel.

Before: FP32 fill before MQA logits

After: fill removed

### Test conditions

| | |
|---|---|
| Hardware | MI355X (gfx950), TP=4 |
| Model | `amd/GLM-5.2-MXFP4`, fp8 KV cache |
| Stack | vLLM `69715823`, AITER `4a1cc77`, flydsl `0.3.0` |
| Server | `--async-scheduling --max-num-seqs 256 --max-num-batched-tokens 16384 --block-size 64 --no-enable-prefix-caching`, `index_topk_freq=4` |
| Env | `VLLM_ROCM_USE_AITER=1`, `VLLM_ROCM_USE_AITER_SPARSE_INDEXER=1`, `AITER_USE_FLYDSL_MOE_SORTING=1`, `VLLM_ROCM_QUICK_REDUCE_QUANTIZATION=INT4` |
| Perf | `vllm bench serve`, random dataset, `--ignore-eos --request-rate inf`, ISL 1024 and 8192, OSL 1024, concurrency 1-256 (16 points) |
| Accuracy | `lm_eval` gsm8k 5-shot chat template with thinking enabled, `max_gen_toks=8192`; RULER `niah_single_2` at 16k |
| Sampling | checkpoint default (temperature 1.0, top_p 0.95), so scores carry roughly +/-0.006 gsm8k and +/-0.02 niah run-to-run |
| Baseline | the same image with only this PR reverted |
| tok/s | aggregate output tokens/s across the 4 GPUs; geomean rows are over the sweep |

### Serving sweep

| ISL | conc | prompts | base tok/s | PR tok/s | &Delta; tput | base TTFT ms | PR TTFT ms | base TPOT ms | PR TPOT ms | &Delta; TPOT |
|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| 1024 | 1 | 16 | 81.3 | 81.2 | **-0.1%** | 1392 | 1387 | 10.80 | 10.82 | +0.2% |
| 1024 | 4 | 32 | 330.5 | 330.5 | **+0.0%** | 131 | 131 | 11.66 | 11.66 | +0.0% |
| 1024 | 8 | 64 | 562.4 | 561.0 | **-0.3%** | 142 | 142 | 13.08 | 13.10 | +0.2% |
| 1024 | 16 | 96 | 875.0 | 880.4 | **+0.6%** | 211 | 211 | 16.44 | 16.35 | -0.5% |
| 1024 | 32 | 128 | 1333.8 | 1333.5 | **-0.0%** | 333 | 333 | 20.58 | 20.56 | -0.1% |
| 1024 | 64 | 256 | 2026.2 | 2042.7 | **+0.8%** | 442 | 438 | 26.96 | 26.66 | -1.1% |
| 1024 | 128 | 384 | 2987.2 | 2985.9 | **-0.0%** | 986 | 979 | 35.19 | 35.17 | -0.1% |
| 1024 | 256 | 768 | 4338.7 | 4337.8 | **-0.0%** | 1630 | 1628 | 50.59 | 50.55 | -0.1% |
| 8192 | 1 | 16 | 84.5 | 84.3 | **-0.3%** | 339 | 336 | 11.50 | 11.53 | +0.3% |
| 8192 | 4 | 32 | 302.2 | 302.4 | **+0.0%** | 447 | 444 | 12.44 | 12.44 | +0.0% |
| 8192 | 8 | 64 | 473.4 | 471.4 | **-0.4%** | 666 | 636 | 15.10 | 15.27 | +1.1% |
| 8192 | 16 | 96 | 665.3 | 668.7 | **+0.5%** | 825 | 826 | 21.44 | 21.33 | -0.5% |
| 8192 | 32 | 128 | 949.8 | 947.2 | **-0.3%** | 1861 | 1770 | 29.40 | 29.59 | +0.6% |
| 8192 | 64 | 256 | 1269.4 | 1270.4 | **+0.1%** | 2991 | 3077 | 44.80 | 44.52 | -0.6% |
| 8192 | 128 | 384 | 1634.2 | 1641.3 | **+0.4%** | 6450 | 6423 | 68.91 | 68.55 | -0.5% |
| 8192 | 256 | 768 | 1909.0 | 1917.4 | **+0.4%** | 14096 | 14072 | 116.63 | 115.82 | -0.7% |

| geomean | tput | TTFT | TPOT |
|---|---:|---:|---:|
| ISL 1024 | **+0.1%** | -0.3% | -0.2% |
| ISL 8192 | **+0.1%** | -1.1% | -0.0% |
| **all 16** | **+0.1%** | -0.7% | -0.1% |

The removed fill is on the prefill path, so the signal lands in TTFT rather than
throughput: **-1.1% geomean at ISL 8192**, reaching -4.9% at concurrency 32
(1861 -> 1770 ms). Throughput and TPOT are flat, as expected for a change worth
0.89% of GPU kernel time in a decode-bound sweep; every case sits within
+/-0.8% on throughput, i.e. inside run-to-run spread.

### Accuracy

| task | baseline | with PR |
|---|---:|---:|
| gsm8k strict-match | 0.9462 | 0.9575 |
| gsm8k flexible-extract | 0.9507 | 0.9621 |
| RULER niah_single_2 @16k | 0.870 | 0.838 |

Unchanged. Both tasks run at the checkpoint's default sampling, so these deltas
are inside run-to-run variation in both directions.

## Tests

```bash
pytest tests/v1/attention/test_rocm_aiter_mla_sparse.py
```

The new test verifies that vLLM forwards `clean_logits=False` to AITER.

### CODE REVIEW

**Comment by amd-sriram** [comment]
@claude review

**Comment by Fangzhou-Ai** [comment]
Can you add some model-level e2e perf (inferenceX 8k1k) and accuracy (gsm8k) test results for verification?

**Comment by amd-sriram** [comment]
> Can you add some model-level e2e perf (inferenceX 8k1k) and accuracy (gsm8k) test results for verification?

@Fangzhou-Ai Now Added the e2e performance and accuracy test results to the description

**Comment by dllehr-amd** [comment]
@amd-sriram  Can you confirm MTP based runs don't need the fill buffer still?  @Rohan138

**Comment by maeehart** [comment]
Commits ee6b15d and cd01bd0 are missing Signed-off-by. Can you rebase with --signoff and push?

**Comment by amd-sriram** [comment]
> @amd-sriram Can you confirm MTP based runs don't need the fill buffer still? @Rohan138

@dllehr-amd @Rohan138 :  MTP runs still don't need the fill. MTP only moves the prefill/decode boundary ([decode_threshold = 1 + speculative tokens](https://github.com/amd-sriram/vllm/blob/b1be181cd0b7f66a94d1ccb8894f01f2d207877d/vllm/v1/attention/ops/rocm_aiter_mla_sparse.py#L1000-L1002)). Anything above it is still a prefill chunk on the same path — and there, producer and consumer get the same bounds tensors: [rocm_fp8_mqa_logits(..., chunk.cu_seqlen_ks, chunk.cu_seqlen_ke)](https://github.com/amd-sriram/vllm/blob/b1be181cd0b7f66a94d1ccb8894f01f2d207877d/vllm/v1/attention/ops/rocm_aiter_mla_sparse.py#L952-L958) then [top_k_per_row_prefill(logits, chunk.cu_seqlen_ks, chunk.cu_seqlen_ke, ...)](https://github.com/amd-sriram/vllm/blob/b1be181cd0b7f66a94d1ccb8894f01f2d207877d/vllm/v1/attention/ops/rocm_aiter_mla_sparse.py#L980-L989). The consumer only scans [ks, ke), exactly what the producer writes, so the out-of-range -inf is dead work. Nothing in that loop is conditioned on spec decode.

Profiled it to be sure : GLM-5.2, {"method": "deepseek_mtp", "num_speculative_tokens": 5}, TP4, 8k/1k, rank 0: 990/990 prefill calls are cp_gather_indexer_k_quant_cache → _gluon_fp8_mqa_logits_kernel → topKPerRowPrefill<512, false>, no exceptions. Still a single consumer under MTP.


mtp_prefill_chain
(Decode side for reference: 62,898 × _gluon_deepgemm_fp8_paged_mqa_logits_preshuffle.kd, untouched by this PR.)

Also — could someone add the ready label?

**Comment by amd-sriram** [comment]
> Commits [ee6b15d](https://github.com/vllm-project/vllm/commit/ee6b15de4fce1472fb63d04d892c127e5f9ccc0f) and [cd01bd0](https://github.com/vllm-project/vllm/commit/cd01bd00e8d5a010eb57304deaeca95296b03db0) are missing Signed-off-by. Can you rebase with --signoff and push?

Done

**Comment by stefanskiasan** [comment]
Independent confirmation on different hardware and a different model, in case
it helps this land: **GLM-5.3 (full, `glm_moe_dsa`), Quark MXFP4 + FP8
attention, MI355X / gfx950, TP4, `max_model_len=1M`, MTP k=4.**

I arrived at the same change independently (as a runtime switch over
`rocm_aiter_mla_sparse.py`) before finding this PR, so the numbers below are
from a separate implementation of the same idea.

## Accuracy — the part I wanted to be sure about

My concern was sequences *shorter* than the indexer's top-k width (2048 here):
if the consumer ever ranked over the unwritten region, it would pick stale
values out of a reused workspace arena rather than garbage, which is the kind
of bug that stays silent. Your reasoning that `top_k_per_row_prefill` only
reads within bounded row ranges is what makes this safe, and it holds
empirically — GSM8K prompts sit well under 2048 tokens, so that benchmark
exercises exactly this case:

| | GSM8K (full 1319) |
|---|---|
| `clean_logits=True` (stock) | 1280 / 1319 = 97.0 % |
| `clean_logits=False` | 1279 / 1319 = 97.0 % |

One question apart, against a standard error of 0.39 % on the full set. No
degradation. (Worth using the full 1319 for this — at 150 questions a
one-question difference is not resolvable, which nearly cost me a working fix
in an unrelated case.)

## Throughput — consistent with your "minimal throughput changes"

Prefill-heavy load, 16k context, 32 concurrent, 128 requests, **caches
disabled** (`--no-enable-prefix-caching`, no LMCache — with the prefix cache on,
this comparison is meaningless):

| | throughput | TTFT |
|---|---|---|
| stock (n=3) | 147.58 tok/s | 19 596 ms |
| `clean_logits=False` (n=6) | 147.06 tok/s | 19 521 ms |

Run-to-run spread is 3.55 tok/s (2.4 %), so a ~0.9 % kernel-time effect is
simply below the resolution of an end-to-end measurement here — which matches
what you report rather than contradicting it. I mention it only so nobody
reads a flat end-to-end number as evidence against the change: the fill
launches do disappear, and removing work that the CUDA path already skips
(`sparse_attn_indexer.py:506,611` and `sparse_attn_indexer_kpool.py:600,911`
all pass `clean_logits=False`) is right on its own terms.

We're holding off on carrying this as a local patch and will pick it up when
this merges. Happy to re-run either measurement on this hardware if that's
useful for review.

**Comment by amd-sriram** [comment]
@stefanskiasan Thanks for the independent confirmation, and Agreed on the full 1319.

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py
Comment (dllehr-amd) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py:776]: Can you confirm if both of these paths work with `clean_logits=false`?

**Review by dllehr-amd — APPROVED** [review]
Approved pending comment that both mqa logit backends in aiter support clean_logits=false

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py
Comment (amd-sriram) [reply vllm/v1/attention/ops/rocm_aiter_mla_sparse.py:776]: The second function supports clean_logits argument - https://github.com/ROCm/aiter/blob/v0.1.22.post1/aiter/ops/triton/attention/fp8_mqa_logits.py#L97. The first function is dead code currently. It existed before - https://github.com/ROCm/aiter/blob/0b3669e5a3bc728bec93366baabd6641235659c0/aiter/ops/triton/fp8_mqa_logits.py. But it got removed as a part of a refactor PR at the beginning of the year - https://github.com/ROCm/aiter/pull/1638. What do you suggest @dllehr-amd, should I delete the first module path?
